xcbgen-rs/src/defs.rs

```diff
@@ -1827,111 +2026,223 @@ impl BuiltInType {
     pub operator: BinaryOperator,
+
+    /// The left hand side of the expression.
     pub lhs: Box<Expression>,
+
+    /// The right hand side of the expression.
     pub rhs: Box<Expression>,
 }
 
+/// The possible binary operators
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 pub enum BinaryOperator {
+    /// Addition
     Add,
+
+    /// Subtraction
     Sub,
+
+    /// Multiplication
     Mul,
+
+    /// (Integer) division
     Div,
+
+    /// Logical and
     And,
+
+    /// A logical left shift
     Shl,
 }
 
+/// An expression with one sub-expression
 #[derive(Debug)]
 pub struct UnaryOpExpr {
+    /// The operator that is applied to the sub-expression.
     pub operator: UnaryOperator,
+
+    /// The sub-expression.
     pub rhs: Box<Expression>,
 }
 
+/// The possible unary operators
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 pub enum UnaryOperator {
+    /// Bitwise negation
     Not,
 }
 
+/// An expression referencing a field from the surrounding context.
 #[derive(Debug)]
 pub struct FieldRefExpr {
+    /// The name of the field that is referenced.
     pub field_name: String,
+
+    /// After resolution, this contains information about the referenced field.
     pub resolved: OnceCell<ResolvedFieldRef>,
 }
 
+/// Information about a field reference.
 #[derive(Debug)]
 pub struct ResolvedFieldRef {
+    /// The kind of reference to another field.
     pub ref_kind: FieldRefKind,
+
+    /// The type of the referenced field.
     pub field_type: TypeRef,
 }
 
+/// Possible kinds of references to fields inside of an expression.
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 pub enum FieldRefKind {
+    /// The reference is to a field of the expression where the field is evaluated.
     LocalField,
+
+    /// The reference is to an external parameter from a surrounding context.
     ExtParam,
+
+    /// The reference is used inside of a [`SumOfExpr`].
     SumOfRef,
 }
 
+/// An expression referencing a parameter from a surrounding context.
 #[derive(Debug)]
 pub struct ParamRefExpr {
+    /// The name of the referenced field.
     pub field_name: String,
+
+    /// The type of the referenced field.
     pub type_: NamedTypeRef,
 }
 
+/// An expression referencing a value from an enumeration.
 #[derive(Debug)]
 pub struct EnumRefExpr {
+    /// The enum that is referenced.
     pub enum_: NamedTypeRef,
+
+    /// The name of the enum's variant that is used.
     pub variant: String,
 }
 
+/// An expression representing the sum over some referenced list.
 #[derive(Debug)]
 pub struct SumOfExpr {
+    /// The name of the list that is used.
     pub field_name: String,
+
+    /// After resolving, a reference to the list that is used.
     pub resolved_field: OnceCell<ResolvedFieldRef>,
+
+    /// The expression that is applied to each element of the list.
+    ///
+    /// If no expression is provided, the elements are summed as-is.
+    // FIXME: Can the Option be replaced with `Expression::ListElementRef`?
```

Comment:
It seems it can be replaced, I'll open a PR for that.

Reply:
Already done in a later commit in this PR.